crypto/x509/v3_purp.c

```diff
@@ -305,44 +305,62 @@ int X509_supported_extension(X509_EXTENSION *ex)
     return 0;
 }
 
-static int setup_dp(X509 *x, DIST_POINT *dp)
+/* return 1 on success, 0 if x is invalid, -1 on (internal) error */
+static int setup_dp(const X509 *x, DIST_POINT *dp)
 {
     const X509_NAME *iname = NULL;
     int i;
 
-    if (dp->reasons) {
+    if (dp->reasons != NULL) {
+        if (dp->distpoint == NULL && sk_GENERAL_NAME_num(dp->CRLissuer) <= 0) {
+            X509err(0, X509_R_INVALID_DISTPOINT);
+            return 0;
+        }
```

Comment:
I'm unsure why you add this inside the `if (dp->reasons != NULL)`. rfc5280 has this text:
```
   While each of these fields is optional, a DistributionPoint MUST NOT
   consist of only the reasons field; either distributionPoint or
   cRLIssuer MUST be present.
```

Which I guess you translated to that. But I think the important part is: "either distributionPoint or cRLIssuer MUST be present.", it does not depend on the reason being present, you always neither at least one of distributionPoint and cRLIssuer.

Reply:
Ah, you are right! Fixed.